src/Diagnostics/single_stack_cache_diagnostics.jl

```diff
@@ -0,0 +1,74 @@
+export single_stack_cache_diagnostics
+using UnPack
+using ..ODESolvers
+using ..Atmos
+using ..Orientations
+using ..SingleStackUtils
+
+"""
+    single_stack_cache_diagnostics(solver_config)
+
+An array of nested NamedTuples, containing results of
+ - `z` - altitude
+ - `prog` - the prognostic state
+ - `aux` - the auxiliary state
+ - `∇flux` - the gradient-flux (diffusive) state
+ - `hyperdiff` - the hyperdiffusive state
+
+and all the nested NamedTuples, merged together,
+from the `precompute` methods.
+"""
+function single_stack_cache_diagnostics(solver_config)
+    bl = solver_config.dg.balance_law
+    t = gettime(solver_config.solver)
+    direction = solver_config.diffdir
+    return [
+        begin
+            @unpack prog, aux, ∇flux, hyperdiff = local_states
+            diffusive = ∇flux
+            state = prog
+            hyperdiffusive = hyperdiff
+
+            _args_fx1 = (; state, aux, t, direction)
+            _args_fx2 = (; state, aux, t, diffusive, hyperdiffusive)
+            _args_src = (; state, aux, t, direction, diffusive)
+
+            cache_fx1 = precompute(bl, _args_fx1, Flux{FirstOrder}())
+            cache_fx2 = precompute(bl, _args_fx2, Flux{SecondOrder}())
+            cache_src = precompute(bl, _args_src, Source())
+
+            # cache_fx1, cache_fx2, and cache_src have overlapping
+            # data, only need one copy, so merge:
+            cache = merge(cache_fx1, cache_fx2, cache_src)
```

Comment:
Maybe better if we introduce a common parent type for `Source` et al. which can be used by a `precompute` implementation that doesn't repeat computations?

Reply:
This would be tricky/involved because the cache would need to persist through `flux_first_order!`, `flux_second_order!` and `source!`. I think repeating the computations is fine since the frequency of the diagnostic is typically much lower than the frequency of the RHS evaluation.